controllers/provider-vsphere/pkg/apis/vsphere/types_cloudprofile.go

```diff
@@ -0,0 +1,179 @@
+
+package vsphere
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CloudProfileConfig contains provider-specific configuration that is embedded into Gardener's `CloudProfile`
+// resource.
+type CloudProfileConfig struct {
+	metav1.TypeMeta
+	// NamePrefix is used for naming NSX-T resources
+	NamePrefix string
+	// Folder is the vSphere folder name to store the cloned machine VM (worker nodes)
+	Folder string
+	// Regions is the specification of regions and zones topology
+	Regions []RegionSpec
+	// DefaultClassStoragePolicyName is the name of the vSphere storage policy to use for the 'default-class' storage class
+	DefaultClassStoragePolicyName string
+	// FailureDomainLabels are the tag categories used for regions and zones.
+	FailureDomainLabels *FailureDomainLabels
+	// DNSServers is a list of IPs of DNS servers used while creating subnets.
+	DNSServers []string
+	// MachineImages is the list of machine images that are understood by the controller. It maps
+	// logical names and versions to provider-specific identifiers.
+	MachineImages []MachineImages
+	// Constraints is an object containing constraints for certain values in the control plane config.
+	Constraints Constraints
+}
+
+// FailureDomainLabels are the tag categories used for regions and zones in vSphere CSI driver and cloud controller.
+// See Cloud Native Storage: Set Up Zones in the vSphere CNS Environment
+// (https://docs.vmware.com/en/VMware-vSphere/6.7/Cloud-Native-Storage/GUID-9BD8CD12-CB24-4DF4-B4F0-A862D0C82C3B.html)
+type FailureDomainLabels struct {
+	// Region is the tag category used for region on vSphere data centers and/or clusters.
+	Region string
+	// Zone is the tag category used for zones on vSphere data centers and/or clusters.
+	Zone string
+}
+
+// RegionSpec specifies the topology of a region and its zones.
+// A region consists of a Vcenter host, transport zone and optionally a data center.
+// A zone in a region consists of a data center (if not specified in the region), a computer cluster,
+// and optionally a resource zone or host system.
+type RegionSpec struct {
+	// Name is the name of the region
+	Name string
+	// VsphereHost is the vSphere host
+	VsphereHost string
+	// VsphereInsecureSSL is a flag if insecure HTTPS is allowed for VsphereHost
+	VsphereInsecureSSL bool
+	// NSXTHost is the NSX-T host
+	NSXTHost string
+	// NSXTInsecureSSL is a flag if insecure HTTPS is allowed for NSXTHost
+	NSXTInsecureSSL bool
+	// TransportZone is the NSX-T transport zone
+	TransportZone string
+	// LogicalTier0Router is the NSX-T logical tier 0 router
+	LogicalTier0Router string
+	// EdgeCluster is the NSX-T edge cluster
+	EdgeCluster string
+	// SNATIPPool is the NSX-T IP pool to allocate the SNAT ip address
+	SNATIPPool string
+
+	// Datacenter is the name of the vSphere data center (data center can either be defined at region or zone level)
+	// +optional
```

Comment:
Remove all `// +optional` tags from this file. Also, ensure that all optional fields are pointer types.